core/src/main/java/com/linecorp/armeria/client/HttpClientPipelineConfigurator.java

```diff
@@ -467,28 +467,51 @@ public void userEventTriggered(ChannelHandlerContext ctx, Object evt) throws Exc
 
         @Override
         public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
-            boolean handled = false;
             if (msg instanceof HttpResponse) {
                 // The server rejected the upgrade request and sent its response in HTTP/1.
                 assert upgradeEvt == UPGRADE_REJECTED;
-                final String connection = ((HttpResponse) msg).headers().get(HttpHeaderNames.CONNECTION);
-                needsToClose = connection != null && Ascii.equalsIgnoreCase("close", connection);
-                handled = true;
+                final HttpResponse res = (HttpResponse) msg;
+                upgradeRejectionCause = "Upgrade request rejected with: " + res.headers();
+                // We can persist connection only when:
+                // - The response has 'Connection: keep-alive' header on HTTP/1.0.
+                // - The response has no 'Connection: close' header on HTTP/1.1.
+                // and:
+                // - The response has 'Content-Length' or 'Transfer-Encoding: chunked',
+                //   i.e. possible to determine the end of the response.
+                //
+                // See: https://www.w3.org/Protocols/rfc2616/rfc2616-sec8.html#sec8.1.2.1
+                //      https://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html#sec4.4
+                needsToClose = !(HttpUtil.isKeepAlive(res) &&
+                                 (HttpUtil.isContentLengthSet(res) ||
+                                  HttpUtil.isTransferEncodingChunked(res)));
+                if (needsToClose) {
+                    // No need to wait till the end of the response.
+                    // Close the connection immediately and finish the upgrade process.
+                    onUpgradeResponse(ctx, false);
+                }
+
+                ReferenceCountUtil.release(msg);
+                return;
+            }
+
+            // We're not going to reuse the connection,
+            // so we just discard everything received.
+            if (needsToClose) {
+                ReferenceCountUtil.release(msg);
+                return;
             }
 
+            // We're not going to close but reuse the connection,
+            // so we wait until the end of the rejecting response.
             if (msg instanceof HttpContent) {
                 if (msg instanceof LastHttpContent) {
-                    // Received the rejecting response completely.
                     onUpgradeResponse(ctx, false);
                 }
-                handled = true;
-            }
-
-            if (!handled) {
-                ctx.fireChannelRead(msg);
-            } else {
                 ReferenceCountUtil.release(msg);
+                return;
             }
+
+            ctx.fireChannelRead(msg);
```

Comment:
question: which one releases the `msg`?

Reply:
The other handlers in the pipeline, or if it reaches at the end of the pipeline, it will be released by the pipeline.

Reply:
Thanks for the explanation. 😄